transport_plugins/jlink/iotile_transport_jlink/jlink.py

```diff
@@ -163,166 +189,195 @@ def _try_connect(self, connection_string):
-        """
-        self._try_connect(connection_string)
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, True)
+        except HardwareError as e:
+            logger.warning("RPC, streaming and tracing won't work, but device still can be flashed or debugged",
+                exc_info=True)
 
-        def _on_finished(_name, control_info, exception):
-            if exception is not None:
-                callback(connection_id, self.id, False, str(exception))
-                return
+        self._connection_id = conn_id
 
-            if control_info is not None:
-                self._control_info = control_info
+    async def disconnect(self, conn_id):
+        self._teardown_connection(conn_id)
 
-            callback(connection_id, self.id, True, None)
-            self._connection_id = connection_id
+        if not self._control_info: # direct debug connection case
+            return
 
-        self._control_thread.command(JLinkControlThread.VERIFY_CONTROL, _on_finished, self._device_info, self._control_info)
+        try:
+            if self.opened_interfaces["streaming"]:
+                await self.close_interface(conn_id, "streaming")
+            if self.opened_interfaces["tracing"]:
+                await self.close_interface(conn_id, "tracing")
+            if self.opened_interfaces["rpc"]:
+                await self.close_interface(conn_id, "rpc")
 
-    def _open_rpc_interface(self, conn_id, callback):
-        """Enable RPC interface for this IOTile device
+            await self._jlink_async.change_state_flag(
+                self._control_info, AsyncJLink.CONNECTION_BIT, False)
+            self.connected = False
+        except pylink.errors.JLinkReadException:
+            logger.warning("Error disconnecting jlink adapter", exc_info=True)
 
-        Args:
-            conn_id (int): the unique identifier for the connection
-            callback (callback): Callback to be called when this command finishes
-                callback(conn_id, adapter_id, success, failure_reason)
-        """
+    async def open_interface(self, conn_id, interface):
+        self._ensure_connection(conn_id, True)
 
-        callback(conn_id, self.id, True, None)
+        if self.opened_interfaces[interface]:
+            return
 
-    def _open_script_interface(self, conn_id, callback):
-        """Enable script streaming interface for this IOTile device
+        self.opened_interfaces[interface] = True
 
-        Args:
-            conn_id (int): the unique identifier for the connection
-            callback (callback): Callback to be called when this command finishes
-                callback(conn_id, adapter_id, success, failure_reason)
-        """
+        if (interface == "tracing" or interface == "streaming") and self._control_info.version < 2:
+            logger.warning("Controller firmware does not support streaming/tracing.")
```

Comment:
Just copying my other comment here from slack:

This should raise a `DeviceAdapterError` saying the interface is not supported on the jlink structure version being used.

Reply:
Fixed.